docs/architecture-decisions/2023-01-12_smart-contract-deployment.md

```diff
@@ -0,0 +1,375 @@
+- Create a contract discovery service API that scans the blockchain transactions ([example](https://developer.algorand.org/articles/developer-preview-of-conduit-a-new-way-to-access-algorand-chain-data/)) to automatically maintain a record of smart contract IDs.
+- Use CI/CD variables ([example](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-output-parameter)) to propagate the deployed app ID to the configuration on the deployment of the client (assuming the same pipeline is used for deployment of smart contract and client(s)).
+- Manually set the application ID via the CI/CD pipeline of the clients with environment-specific values (that are passed in via environment variables) and change them to reflect any changes to the smart contracts (note: this isn't sustainable for frequently changing contracts, but is a simple method for infrequently changing contracts).
+
+In order to provide the simplest implementation possible to handle this complexity, AlgoKit v1 will:
+
+- Add payloads to application creation transactions to allow them to be resolved by name for a known creator account using the indexer.
+- Allow application IDs to be resolved by name using environment variables for situations where a developer is using ephemeral creator accounts.
+- On-chain resolution will be left for developers to implement (likely by hardcoding or Global-state-based configurable application IDs).
+
+More sophisticated options could potentially by implemented in the future.
+
+### Automated vs manual deployments
+
+In order to support Continuous Delivery for smart contracts there is a need to support automated deployments of smart contracts via deployment pipeline.
+
+This presents a number of challenges to solve for though:
+
+- The deployment pipeline needs to determine if the smart contract already exists or not in the network being deployed to
+- Depending on whether the contract is immutable and/or permanent and if it already exists or not in the network being deployed to, the pipeline needs to handle upgrade, deletion and/or creation of the contract (and this behaviour may need to be switched depending on the network being deployed to)
+- In order to protect integrity of MainNet contracts the deployment process should probably have some safeguards for MainNet that by default prevent destructive operations and require some kind of human opt-in to perform those operations
+- Exposing private key mnemonics for MainNet necessitates they are "hot wallets" which may be undesirable for privileged accounts responsible for high value smart contracts (e.g. smart contracts locking $m's or $b's), alternatively low privilege scenarios may benefit from (encrypted secret) mnemonic storage to improve developer experience and operational overhead
+- Supporting cold storage wallets for high privilege accounts, without resorting the manual deployments (which present their own risks) is tricky
```

Comment:
I expect that deployment often just consists, at the end, of a group of transactions to send to the blockchain (maybe multiple groups).
One solution in that case is to output the unsigned transaction file, ask the user to take it and sign it and give it back, then send it.

But this would not be in a CI/CD pipeline, rather just a command line deployment process (`algokit deploy mainnet`).